EL-3599 - typeahead documentation

TaraTodd123 opened the pull request:
#### Checklist
* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licenced under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-3599

#### Description of Proposed Changes

- Removed typeahead documentation from tag section
- Typeahead is now in its own section in Input Controls
- Added additional documentation for <code>active</code>, <code>loading</code>, <code>onFilterChange</code> and <code>selectOnEnter</code>
- Example added to Typeahead Section and Plunker
- Added link in tag input to typeahead and vice versa.

#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-3599-typehead-documentation

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,55 @@
+    /** Load the options and filter the them */
+    loadOptions(pageNum: number, pageSize: number, filter: string) {
+
+        const displayValues = ['Alpha', 'Beta', 'Gamma', 'Delta', 'Epsilon', 'Zeta', 'Eta', 'Theta', 'Iota', 'Kappa', 'Lambda', 'Mu', 'Nu', 'Xi', 'Omicron', 'Pi', 'Rho', 'Sigma', 'Tau', 'Upsilon', 'Phi', 'Chi', 'Psi', 'Omega'];
```
We should probably make these items a class value, eg:

![image](https://user-images.githubusercontent.com/20795331/62936529-f51a7380-bdc1-11e9-93ee-ed19da71a974.png)

You would need to add a reference to a function bound to the correct context, otherwise you'll get an error saying `this._values` is undefined.

It would probably good to make a longer list to demonstrate paging. Perhaps also add a small delay to simulate server loading time. Consider using `chance.js` to populate the list with lots of data for paging.

TaraTodd123 replied:
Added <code>chance.js</code>

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.html:
```diff
@@ -0,0 +1,48 @@
+                   $event.stopPropagation();">
+            <ux-typeahead #typeahead
+                          class="typeahead-example"
+                          [open]="dropdownOpened"
```
Consider renaming to `dropdownOpen` which indicates its current open state, not if it has been opened at some point

TaraTodd123 replied:
renamed

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.css:
```diff
@@ -0,0 +1,4 @@
+.typeahead-example {
```
changed

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/typeahead.component.html:
```diff
@@ -0,0 +1,188 @@
+    It has the following properties:</p>
+
+<uxd-api-properties tableTitle="Inputs">
+    <tr uxd-api-property name="options" type="	any[] | InfiniteScrollLoadFunction">
```
Changed to type <code>T</code>

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.html:
```diff
@@ -0,0 +1,48 @@
+    </div>
+</div>
+
+<div class="row uxd-customize-example">
```
Removed

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.html:
```diff
@@ -0,0 +1,48 @@
+                   (click)="dropdown(); $event.stopPropagation();"
+                   (keydown)="dropdown(); typeaheadKeyService.handleKey($event, typeahead);
+                   $event.stopPropagation();">
+            <ux-typeahead #typeahead
```
Added so escape and clicking out side closes dropdown

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    /** Opening the typeahead dropdown with filter options. */
+    dropdown() {
```
Removed

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,55 @@
+    /** Load the options and filter the them */
+    loadOptions(pageNum: number, pageSize: number, filter: string) {
+
+        const displayValues = ['Alpha', 'Beta', 'Gamma', 'Delta', 'Epsilon', 'Zeta', 'Eta', 'Theta', 'Iota', 'Kappa', 'Lambda', 'Mu', 'Nu', 'Xi', 'Omicron', 'Pi', 'Rho', 'Sigma', 'Tau', 'Upsilon', 'Phi', 'Chi', 'Psi', 'Omega'];
```
Removed with using <code>chance.js</code>

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,55 @@
+    private  _value$ = new BehaviorSubject<string>(null);
+
+    /** Load the options and filter the them */
+    loadOptions(pageNum: number, pageSize: number, filter: string) {
```
Added return types

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    /** selected typeahead option and closing dropdown **/
+    singleOptionSelected(event: TypeaheadOptionEvent): void {
```
renamed

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,55 @@
+
+    /** The selected option in input field */
+    set value(value: string) {
+        this._value$.next(value);
```
removed

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/typeahead.component.html:
```diff
@@ -0,0 +1,188 @@
+                   class="form-control"
+                   (keydown.enter)="$event.preventDefault()"
+                   (click)="dropdown(); $event.stopPropagation();"
+                   (keydown)="dropdown(); typeaheadKeyService.handleKey($event, typeahead);
```
removed

TaraTodd123 commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.html:
```diff
@@ -0,0 +1,48 @@
+<div class="row">
+    <div class="col-md-12 form-group-validate has-feedback">
```
removed

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,54 @@
+})
+export class AppComponent {
+
+    tagDocumentationRoute: string;
```
This variable is only used in our docs site so doesn't need to be included in the sample code

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,54 @@
+export class AppComponent {
+
+    tagDocumentationRoute: string;
+    values: string[] = [];
```
When using arrays in HTML we always try to use `ReadonlyArray` instead of a normal array. This forces the reference to change each time the array changes and can improve performance in large applications so we should promote this good practice in our sample code.

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.html:
```diff
@@ -0,0 +1,27 @@
+                          [filter]="input"
+                          [options]="loadOptionsFn"
+                          [openOnFilterChange]="true"
+                          [pageSize]="5"
```
We should make the default page size larger, currently there is a flicker when the typeahead opens and the scrollbar jumps several times as it has to load multiple pages to fill the height of the typeahead.

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.html:
```diff
@@ -0,0 +1,27 @@
+                   (click)="dropdownOpen = true; $event.stopPropagation();"
+                   (keydown.escape)="dropdownOpen = false; $event.stopPropagation();"
+                   [attr.aria-controls]="typeahead">
+            <ux-typeahead #typeahead
```
Arrow keys should allow navigation of the typeahead items. We have a typeahead key service to handle this, we should use it here.

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.html:
```diff
@@ -0,0 +1,27 @@
+                   [placeholder]="'Enter Text'"
+                   class="form-control"
+                   (keydown.enter)="dropdownOpen = true; $event.preventDefault()"
+                   (click)="dropdownOpen = true; $event.stopPropagation();"
```
Do we need these `stopPropgation` and `preventDefault` calls? We might do, taking them out I didn't see any difference but I could be wrong

TaraTodd123 commented:
- Added <code>uxClickOutside</code>
- Removed unused code and imports from plunker
- Changes typeahead value type to <code>ReadOnlyArray</code>
- Added  <code>TypeaheadKeyService</code>

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,45 @@
+
+    /** Load the options and filter the them */
+    loadOptions(pageNum: number, pageSize: number, filter: string): ReadonlyArray<string> {
+        return this.values.filter(tag => tag.toLowerCase().indexOf(filter.toLowerCase()) !== -1).slice(pageNum * pageSize, (pageNum + 1) * pageSize);
```
This line is causing scrollbars to appear on the sample code snippet. Split this over multiple lines, eg:

```typescript
return this.values.filter(tag => tag.toLowerCase().indexOf(filter.toLowerCase()) !== -1)
    .slice(pageNum * pageSize, (pageNum + 1) * pageSize);
```

ashley-hunter replied:
I also just noticed one original review comment has been missed, could we add a 1 or 2 second delay here to simulate page loading from the server rather than having pages loading immediately. Might also be good to bump the number of items up from 40 to like 200 to give a good preview of what it is like to have multiple pages.

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/typeahead.component.html:
```diff
@@ -0,0 +1,191 @@
+<div class="row">
+    <div class="col-md-12 has-feedback">
+        <form class="input-typeahead-form">
+            <input [value]="value"
```
I dont think we need `value` here? `ngModel` should sync value

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/typeahead.component.html:
```diff
@@ -0,0 +1,191 @@
+                   [ngModelOptions]="{standalone: true}"
+                   placeholder="Enter Text"
+                   class="form-control"
+                   (keydown.enter)="dropdownOpen = true;"
```
If we look at the tag input component, enter does not open the typeahead, instead the down arrow does, so perhaps change to use it instead.

TaraTodd123 commented:
- Text changes for typeahead dcoumentation
- removed <code>[value]</code>
- Delay added for page loading
- Updates to component: not setting open to true if filter value and selected value match
- selecting first to highlighted if paging

donna-marijne commented on docs/app/pages/components/components-sections/input-controls/typeahead/typeahead.component.html:
```diff
@@ -0,0 +1,190 @@
+
+<hr>
+
+<p>The ux-typeahead component can be used to apply a dropdown typeahead list to a text input. To match the
```
`code` around `ux-typeahead` please, and split this into two paragraphs as mentioned.

donna-marijne commented on docs/app/pages/components/components-sections/input-controls/typeahead/typeahead.component.html:
```diff
@@ -0,0 +1,190 @@
+
+<p>The ux-typeahead component can be used to apply a dropdown typeahead list to a text input. To match the
+    width of the input control, place both inside a container with a defined width and <code>position:
+        relative.</code></p>
```
The period "." is not part of the `code`.

TaraTodd123 commented:
Sorry about that, Updated now.

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/snippets/app.ts:
```diff
@@ -0,0 +1,44 @@
+    constructor(public typeaheadKeyService: TypeaheadKeyService<string>) {
+
+        /* Adding values to typeahead list */
+        for (let index = 0; index < 40; index++) {
```
I think this needs to be bumped up to 200 as well

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/typeahead.component.ts:
```diff
@@ -0,0 +1,69 @@
+    selector: 'uxd-components-typeahead',
+    templateUrl: 'typeahead.component.html',
+    styleUrls: ['./typeahead.component.less'],
+    encapsulation: ViewEncapsulation.None
```
Can this be removed? Looks like there is only one style in the stylesheet that only effects this component so I don't think this is needed anymore.

ashley-hunter commented on src/components/typeahead/typeahead.component.ts:
```diff
@@ -350,9 +354,7 @@ export class TypeaheadComponent<T = any> implements OnChanges, OnDestroy {
-        // Clear previous highlight
-        this.highlighted$.next(null);
-        if (this.selectFirst && this.options) {
+        if (this.selectFirst && this.options && this._currentPageNumber === 0) {
```
I think instead of having to add the `_currentPageNumber` variable, the `loaded` event which calls this function emits an object the contains the page that was loaded, we can us it instead.

TaraTodd123 commented:
Above 2 fixed

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/typeahead.component.html:
```diff
@@ -0,0 +1,190 @@
+<hr>
+
+<p>The <code>ux-typeahead</code> component can be used to apply a dropdown typeahead list to a text input.</p>
+<p>To match the width of the input control, place both inside a container with a defined width and
```
I don't think needing a `defined width` is accurate as our example does not have a defined with. I think the only important part here is that the container element has `position: relative`.

ashley-hunter commented on docs/app/pages/components/components-sections/input-controls/typeahead/typeahead.component.html:
```diff
@@ -0,0 +1,190 @@
+<p>The <code>ux-typeahead</code> component can be used to apply a dropdown typeahead list to a text input.</p>
+<p>To match the width of the input control, place both inside a container with a defined width and
+    <code>position: relative</code>.</p>
+<p>It has the following properties:</p>
```
Perhaps to keep this better aligned with our other documentation sections this could be something like:

> The following properties can be used to customize the appearance and behavior of the `ux-typeahead` component:

TaraTodd123 commented:
Text updates added
